tpm2/direct/reflect.go

```diff
@@ -0,0 +1,999 @@
+	cc := cmd.Command()
+	if rsp.Response() != cc {
+		return fmt.Errorf("cmd and rsp must be for same command: %v != %v", cc, rsp.Response())
+	}
+	sess, err := cmdAuths(cmd)
+	if err != nil {
+		return err
+	}
+	sess = append(sess, extraSess...)
+	if len(sess) > 3 {
+		return fmt.Errorf("too many sessions: %v", len(sess))
+	}
+	hasSessions := len(sess) > 0
+	// Initialize the sessions, if needed
+	for i, s := range sess {
+		if err := s.Init(t); err != nil {
+			return fmt.Errorf("initializing session %d: %w", i, err)
+		}
+		if err := s.NewNonceCaller(); err != nil {
+			return err
+		}
+	}
+	handles := cmdHandles(cmd)
+	names, err := cmdNames(cmd)
+	if err != nil {
+		return err
+	}
+	parms, err := cmdParameters(cmd, sess)
+	if err != nil {
+		return err
+	}
+	sessions, err := cmdSessions(t, sess, cc, names, parms)
+	if err != nil {
+		return err
+	}
+	hdr := cmdHeader(hasSessions, 10 /* size of command header */ +len(handles)+len(sessions)+len(parms), cc)
+	command := append(hdr, handles...)
+	command = append(command, sessions...)
+	command = append(command, parms...)
+
+	// Send the command via the transport.
+	response, err := tpmutil.RunCommandRaw(t.transport, command)
+	if err != nil {
+		return err
+	}
+
+	// Parse the command directly into the response structure.
+	rspBuf := bytes.NewBuffer(response)
+	err = rspHeader(rspBuf)
+	if err != nil {
+		var bonusErrs []string
+		// Emergency cleanup, then return.
+		for _, s := range sess {
+			if err := s.CleanupFailure(t); err != nil {
+				bonusErrs = append(bonusErrs, err.Error())
+			}
+		}
+		if len(bonusErrs) != 0 {
+			return fmt.Errorf("%w - additional errors encountered during cleanup: %v", err, strings.Join(bonusErrs, ", "))
+		}
+		return err
+	}
+	err = rspHandles(rspBuf, rsp)
+	if err != nil {
+		return err
+	}
+	rspParms, err := rspParametersArea(hasSessions, rspBuf)
+	if err != nil {
+		return err
+	}
+	if hasSessions {
+		// We don't need the TPM RC here because we would have errored
+		// out from rspHeader
+		// TODO: Authenticate the error code with sessions, if desired.
+		err = rspSessions(rspBuf, TPMRCSuccess, cc, names, rspParms, sess)
+		if err != nil {
+			return err
+		}
+	}
+	err = rspParameters(rspParms, sess, rsp)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Marshal will serialize the given values, appending them onto the given writer.
+// Returns an error if any of the values are not marshallable.
+func Marshal(w io.Writer, vs ...interface{}) error {
+	var reflects []reflect.Value
+	for _, v := range vs {
+		reflects = append(reflects, reflect.ValueOf(v))
+	}
+	var buf bytes.Buffer
+	if err := marshal(&buf, reflects...); err != nil {
+		return err
+	}
+	_, err := io.Copy(w, &buf)
+	return err
+}
+
+// marshal will serialize the given values, appending them onto the given
+// buffer.
+// Returns an error if any of the values are not marshallable.
+func marshal(buf *bytes.Buffer, vs ...reflect.Value) error {
```

Comment:
Consider changing buf to an io.Writer or taking advantage of the properties of bytes.Buffer (no errors, seekable, etc)

Reply:
We do, in some places (buf.Len()). For consistency, seems fine for all the internal marshalling code to pass pointers to bytes.Buffer.